### PR TITLE
Feature: Form inherits campaign goal

### PR DESCRIPTION
**Pull request by alaca**
## Description

WIP

This PR adds functionality that sets the form goal based on the campaign goal. 

## Affects

Form goal

## Testing Instructions

Create a new campaign
Set campaign goal
Verify that the default campaign form has the same goal settings as the campaign
Create another form and verify that the new form also has the same goal settings as the campaign. 

## Pre-review Checklist

-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php [hunk 1]
```diff
@@ -50,6 +52,7 @@ class DonationFormGoalData implements Arrayable
     public $goalEndDate;
 
     /**
+     * @unreleased Campaign is the default goal type
```
Comment (glaubersilva) [inline src/DonationForms/DataTransferObjects/DonationFormGoalData.php:55]: Can we remove this? I think you rolled back the changes previously made in the constructor but kept the `@unreleased` tag.

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php [hunk 2]
```diff
@@ -77,21 +80,15 @@ public function getCurrentAmount()
...
+            case GoalType::DONORS_FROM_SUBSCRIPTIONS():
                 return $query->countDonors();
             case GoalType::DONATIONS():
-                return $query->count();
             case GoalType::SUBSCRIPTIONS():
                 return $query->count();
             case GoalType::AMOUNT_FROM_SUBSCRIPTIONS():
                 return $query->sumInitialAmount();
-            case GoalType::DONORS_FROM_SUBSCRIPTIONS():
-                return $query->countDonors();
```
Comment (glaubersilva) [inline src/DonationForms/DataTransferObjects/DonationFormGoalData.php:94]: Let's add an `@unreleased` tag for these changes.

File: src/FormBuilder/Routes/CreateFormRoute.php [hunk 3]
```diff
@@ -50,4 +48,29 @@ public function __invoke()
...
+                    'goalType' => $campaign->goalType->getValue(),
+                ];
+            }
+        }
+
+        return [
+            'enableDonationGoal' => true,
+            'goalAmount' => 1000,
+        ];
+    }
```
Comment (glaubersilva) [inline src/FormBuilder/Routes/CreateFormRoute.php:75]: The best approach here would be to rollback these changes and implement this logic using hooks inside the `Campaigns` domain. You can use the `givewp_donation_form_creating` for that. 

As a sample, you can take a look at the `AddCampaignFormFromRequest` class which is triggered on the `givewp_donation_form_created` action.

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 namespace Give\DonationForms\DataTransferObjects;
 
+use Give\Campaigns\CampaignDonationQuery;
+use Give\Campaigns\Models\Campaign;
 use Give\DonationForms\DonationQuery;
 use Give\DonationForms\Models\DonationForm;
 use Give\DonationForms\Properties\FormSettings;
```
Comment (glaubersilva) [inline src/DonationForms/DataTransferObjects/DonationFormGoalData.php:9]: Let's remove all unused imports from this file.

image

File: src/Campaigns/Repositories/CampaignRepository.php [hunk 5]
```diff
@@ -38,6 +38,17 @@ public function getById(int $id)
             ->get();
     }
 
+    /**
+     * Get Campaign by Form ID
+     */
+    public function getByFormId(int $id)
```
Comment (glaubersilva) [inline src/Campaigns/Repositories/CampaignRepository.php:44]: Let's add an `@unreleased` tag here. Also, I added a method with the same name on #7571 so we'll need to manage this conflict later.

File: src/Campaigns/Models/Campaign.php [hunk 6]
```diff
@@ -110,6 +110,16 @@ public static function find($id): ?Campaign
         return give(CampaignRepository::class)->getById($id);
     }
 
+    /**
+     * Find campaign by Form ID
+     *
+     * @unreleased
+     */
+    public static function findByFormId($id): ?Campaign
```
Comment (glaubersilva) [inline src/Campaigns/Models/Campaign.php:118]: Same thing here, I added a method with the same name on #7571 so we'll need to manage this conflict later.

File: src/FormBuilder/Routes/CreateFormRoute.php (hunk 3 above)
Comment (glaubersilva) [reply src/FormBuilder/Routes/CreateFormRoute.php:75]: Another thing to consider is that we also need to use the `save_post_give_forms` hook to add this same logic for v2 forms. 

I used this same hook on #7571 to implement a way to add v2 forms to the campaigns. I did that by changing the `AddCampaignFormFromRequest` class to this: https://github.com/impress-org/givewp/blob/feature/feature-flag-GIVE-1264/src/Campaigns/Actions/AddCampaignFormFromRequest.php#L16-L26

**Review by glaubersilva — CHANGES_REQUESTED** [review]
@alaca I left a few comments with improvement suggestions. 

Also, after pulling this branch I could see the campaign on the list table and I got the following fatal error in the logs:

```
[18-Oct-2024 17:46:55 UTC] PHP Fatal error:  Uncaught TypeError: Return value of Give\Campaigns\DataTransferObjects\CampaignGoalData::getPercentage() must be of the type int, float returned in C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Campaigns\DataTransferObjects\CampaignGoalData.php:83
Stack trace:
#0 C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Campaigns\DataTransferObjects\CampaignGoalData.php(53): Give\Campaigns\DataTransferObjects\CampaignGoalData->getPercentage()
#1 C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Campaigns\ListTable\Columns\GoalColumn.php(39): Give\Campaigns\DataTransferObjects\CampaignGoalData->__construct(Object(Give\Campaigns\Models\Campaign))
#2 C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Framework\ListTable\ListTable.php(135): Give\Campaigns\ListTable\Columns\GoalColumn->getCellValue(Object(Give\Campaigns\Models\Campaign), 'en-US')
#3 C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Framewo in C:\Users\glaub\Local Sites\givewp\app\public\wp-content\plugins\give\src\Campaigns\DataTransferObjects\CampaignGoalData.php on line 83
```
image

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php (hunk 2 above)
Comment (alaca) [reply src/DonationForms/DataTransferObjects/DonationFormGoalData.php:94]: For what? I just grouped the related cases.

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php (hunk 1 above)
Comment (alaca) [reply src/DonationForms/DataTransferObjects/DonationFormGoalData.php]: Resolved https://github.com/impress-org/givewp/pull/7569/commits/9d80e7d9e7b3c3415aac2fc112d2113d0e6773d7

File: src/FormBuilder/Routes/CreateFormRoute.php (hunk 3 above)
Comment (alaca) [reply src/FormBuilder/Routes/CreateFormRoute.php]: Resolved https://github.com/impress-org/givewp/pull/7569/commits/7acbe0937fcfd1bf7d67f005a807158f172aa187

File: src/DonationForms/DataTransferObjects/DonationFormGoalData.php (hunk 4 above)
Comment (alaca) [reply src/DonationForms/DataTransferObjects/DonationFormGoalData.php:6]: Resolved https://github.com/impress-org/givewp/pull/7569/commits/4ca37fe73d1c381bb8974334b0fec35f9d9386b0

File: src/Campaigns/Repositories/CampaignRepository.php (hunk 5 above)
Comment (alaca) [reply src/Campaigns/Repositories/CampaignRepository.php:48]: Resolved https://github.com/impress-org/givewp/pull/7569/commits/cb91501f43449a1ffadc47b7ead992bd08e1a890

**Review by glaubersilva — APPROVED** [review]
👍

**Pull request merged**